### PR TITLE
BM - Implement Soft Deletes

### PR DESCRIPTION
**Pull request by BreeMadison**
## Why

[Monday.com Ticket](https://code4community-team.monday.com/boards/702429655/pulses/742442213)

## This PR

- Added columns `deleted` and `deleted_timestamp` to the Users and Team tables

- Removed hard delete functionality from the protected user processor and team processor implementation 

- Replaced the above with soft delete functionality, where the deleted entity's deleted value is switched to TRUE, and the deleted timestamp is updated to the current time

## Verification Steps

Be sure to run mvn clean install to update database

For User
1. Create a test user (either by hitting the API or in the database
2. Delete them by hitting the API
3. Check the database and make sure their deleted column is set to true and that the timestamp is the current time
4. If the user was a team leader make sure the team they led also has it's deleted column set to true and that it's deleted timestamp is current

For Team
1. Create a test user 
2. Using that user make a new team by hitting the API
3. Verify that the team is in the database
4. Delete the team by hitting the API's protected teams route
5. Verify that within the database that team's deleted column is set to true and that the timestamp is the current time

### CODE REVIEW

File: persist/src/main/resources/db/migration/V5__SoftDeletes.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+ALTER TABLE users
+ADD COLUMN deleted BOOLEAN NOT NULL DEFAULT FALSE;
+ALTER TABLE users
+    ADD COLUMN deleted_timestamp timestamp;
```
Comment (jackblanc) [inline persist/src/main/resources/db/migration/V5__SoftDeletes.sql:4]: ⛏️  Can these be combined to one statement per table?

**Review by jackblanc — COMMENTED** [review]
Looks great, but I'm not seeing how we prevent returning these "deleted" rows when selecting data from the DB

**Comment by BreeMadison** [comment]
> Looks great, but I'm not seeing how we prevent returning these "deleted" rows when selecting data from the DB

@jackblanc Yeah, I wasn't sure how we wanted to handle that since the ticket didn't give much description other than just flipping a true/false value when a team or user got deleted. But, I was thinking that we can throw a userDoesNotExistException if the code tries to do an operation on a user or team who is "deleted". If you have any other suggestions I'm happy to hear them

File: persist/src/main/resources/db/migration/V5__SoftDeletes.sql (hunk 1 above)
Comment (BreeMadison) [reply persist/src/main/resources/db/migration/V5__SoftDeletes.sql:4]: I tried to put them into one statement per table, but I ran into this error: 
` Failed to execute goal org.jooq:jooq-codegen-maven:3.12.0:generate (jooq-codegen) on project persist: Error running jOOQ code generation tool: Unexpected token or missing query delimiter: [2:50] ...
[ERROR] ADD COLUMN deleted BOOLEAN NOT NULL DEFAULT FALSE[*],
[ERROR] ADD COLUMN deleted_timestamp timestamp;
[ERROR] ALTER TABLE team
`

After doing some research, it seems that one of the versions of our dependencies does not allow for multiple add column statements within an alter table statement

**Comment by rprende** [comment]
> > Looks great, but I'm not seeing how we prevent returning these "deleted" rows when selecting data from the DB
> 
> @jackblanc Yeah, I wasn't sure how we wanted to handle that since the ticket didn't give much description other than just flipping a true/false value when a team or user got deleted. But, I was thinking that we can throw a userDoesNotExistException if the code tries to do an operation on a user or team who is "deleted". If you have any other suggestions I'm happy to hear them

@BreeMadison any updates? if not, we can connect about specifics for this ticket

**Comment by BreeMadison** [comment]
> > > Looks great, but I'm not seeing how we prevent returning these "deleted" rows when selecting data from the DB
> 
> > 
> 
> > @jackblanc Yeah, I wasn't sure how we wanted to handle that since the ticket didn't give much description other than just flipping a true/false value when a team or user got deleted. But, I was thinking that we can throw a userDoesNotExistException if the code tries to do an operation on a user or team who is "deleted". If you have any other suggestions I'm happy to hear them
> 
> 
> 
> @BreeMadison any updates? if not, we can connect about specifics for this ticket

Can we connect possibly later this week? I just want to make sure there's no possibility of a leaked deleted row being returned from the our api @rprende

**Pull request closed without merging**